esp-hal/src/sync.rs

```diff
@@ -1,23 +1,38 @@
+//! Under construction: This is public only for tests, please avoid using it.
+
+#[cfg(single_core)]
+use core::cell::Cell;
 use core::cell::UnsafeCell;
 
 mod single_core {
+    use core::sync::atomic::{compiler_fence, Ordering};
+
     pub unsafe fn disable_interrupts() -> critical_section::RawRestoreState {
         cfg_if::cfg_if! {
             if #[cfg(riscv)] {
                 let mut mstatus = 0u32;
                 core::arch::asm!("csrrci {0}, mstatus, 8", inout(reg) mstatus);
-                ((mstatus & 0b1000) != 0) as critical_section::RawRestoreState
+                let token = ((mstatus & 0b1000) != 0) as critical_section::RawRestoreState;
             } else if #[cfg(xtensa)] {
                 let token: critical_section::RawRestoreState;
                 core::arch::asm!("rsil {0}, 5", out(reg) token);
-                token
             } else {
                 compile_error!("Unsupported architecture")
             }
-        }
+        };
+
+        // Ensure no subsequent memory accesses are reordered to before interrupts are
+        // disabled.
+        compiler_fence(Ordering::SeqCst);
```

Comment:
```suggestion
        compiler_fence(Ordering::Acquire);
```

Reply:
Nah, if the cortex-m folks used seqcst, I'll do, too. The critical-section crate needs "at least" acquire/release ordering.